2 to 10 times faster C++ core libraries compilation & rebuild.

deathbaba opened the pull request:
More details here: http://blog.rburchell.com/2013/10/every-time-you-configordered-kitten-dies.html

Before changes (in master), Debug full project rebuild on Mac with make -j8:
real	4m44.620s
user	17m44.706s
sys	1m22.947s

After changes in this branch, Debug full project rebuild on Mac with make -j8:
real	3m23.325s
user	20m33.222s
sys	1m38.610s

Before changes (in master), Debug build after indexer/feature.cpp modification:
real	0m36.891s
user	0m32.087s
sys	0m6.036s

After changes in this branch, Debug build after indexer/feature.cpp modification:
real	0m11.101s
user	0m40.733s
sys	0m5.901s

mapsmetest commented:
Jenkins asks: can anyone have a look at this request?
If you want Jenkins to compile and test this, write "test this please" or JT_TP (without the underscore in the middle).

Zverik commented on omim.pro:
```diff
@@ -26,75 +29,139 @@ win32:CONFIG(drape) {
-SUBDIRS = 3party base geometry coding
-SUBDIRS += indexer
-SUBDIRS += routing
+SUBDIRS = 3party base coding geometry indexer routing
```
Ты же написал чуть выше, что есть зависимость между indexer и geometry?

deathbaba replied:
Написал, тут это не важно. Порядок после убирания ordered стал не важен.

Zverik commented on omim.pro:
```diff
@@ -26,75 +29,139 @@ win32:CONFIG(drape) {
-  CONFIG(desktop) {
-    SUBDIRS += qt
-  }
+  SUBDIRS *= anim graphics gui render search map
```
А для этого всего не нужно зависимости проставлять?

deathbaba replied:
Статические библиотеки собираются независимо и параллельно, чтобы сделать архив - компилятору не нужны остальные библиотеки. Но если библиотека использует код другой библиотеки, тогда любой бинарник, который ее использует, должен так же подключить и другие библиотеки тоже.

Zverik commented:
JTTP

Zverik commented:
Я проверю в понедельник, как это на всём собирается. Подозреваю, тут где-то пропущены зависимости, но обосновать не могу.

syershov commented on common.pri:
```diff
@@ -23,15 +23,12 @@ VERSION = $${VERSION_MAJOR}.$${VERSION_MINOR}
-# Do not remove depend_includepath in CONFIG. Need to check modifications within ROOT_DIR.
-#CONFIG -= depend_includepath
+# Turns off any external headers modification check.
+CONFIG -= depend_includepath
```
http://blog.mgsxx.com/?p=2070

DEPENDPATH

Документация: “Эта переменная содержит список всех директорий для поиска в них зависимостей. Она используется при прохождении по включенным (include) файлам”. Хрен поймешь, что это такое и нафик оно нужно. Как я выяснил, это еще и не совсем верная и изрядно устаревшая информация.

Задумка этой переменной в следующем: быть подмножеством INCLUDEPATH. Только включенные файлы из каталогов этого подмножества создают зависимости (т.е. прописываются как предпосылки к файлам, их включающим прямо или косвенно). Таким образом можно было избежать включения в список зависимостей стабильных файлов (заголовочные файлы стандартной библиотеки C++ и Qt, к примеру). Такая оптимизация значительно сокращает работу для make, т.к. гораздо меньше файлов нужно проверять на “свежесть”.

Но практика показала, что от такой оптимизации больше вреда, чем пользы. Люди должны были руками за этой переменной следить, чтобы не напороться на трудноотслеживаемые глюки. Учитывая, что любую документацию никто не читает, а в доке по qmake к тому же ничего толком и не написано, то именно после такого столкновения с глюками люди об этой переменной и узнавали. Полнейшее безобразие. Так что, в конце концов, разработчики решили убрать этот механизм под ковер. В Qt 4 этого сделать было нельзя, не поломав старый код, поэтому ограничились введением способа сделать так, чтобы DEPENDPATH всегда был равен INCLUDEPATH. Делается это с помощью CONFIG += depend_includepath. В Qt 5 эта настройка включена по умолчанию. Мазохисты могут ее отключить (CONFIG -= depend_includepath), но я не советую. Жизнь слишком коротка, чтобы тратить ее на всякие глупости. Из глупостей: до 5 версии Qt DEPENDPATH по неведомым причинам включался в VPATH, что приводило к хитрым глюкам.

Резюме: если включена опция depend_includepath, то все включенные файлы будут создавать зависимости, причем рекурсивно. Таким образом, автоматически обеспечивается, что изменение любого заголовочного файла, включенного откуда угодно каким угодно путем, приведет к правильной перекомпиляции при вызове make. Аминь.

deathbaba replied:
Да, все мы читали эту статью. И именно после ее прочтения, сжав зубы, решили замедлить компиляцию в разы, закомментировав опцию.
П проблемы с этой опцией у нас возникали из-за того, что наши инклуды из папки std не были добавлены в проект, и, соответственно, после любых в них изменений (например, активный переход на C++11) проектные файлы не перекомпилировались.

Уже задолго после комментирования строки с отключением опции (и замедления сборки) я добавил все инклуды из папки std в common.pri - т.е. сейчас любые изменения в них будут корректно приводить к перекомпиляции нашего проекта.

Поэтому сейчас можно снова раскомментировать строку, которая отключит проверку на изменение системных инклудов (мы же не меняем системные инклуды при разработке, верно?) и наслаждаться супер быстрой сборкой проекта.

Напомню, что мы прекрасно работали уже с этой опцией больше года, после выхода Qt5, до тех пор, пока не столкнулись с проблемой перекомпиляции std инклудов, описанной выше.

Если все наши хидеры в проекте будут присутствовать в .pro файлах, то никаких проблем не будет. Если кто-то забудет добавить хидер в .pro файл - тогда, при модификации такого “анонимного” хидера qmake не определит, что он изменился, и не пересоберет проект. Риск того, что разработчик забудет добавить хидер в .pro файл, несомненно есть. Но он небольшой, понятно как исправляется (можно даже скриптом это узнать и исправить), а польза от быстрой сборки - огромная. Особенно под iOS или Android, где собираются по две-три архитектуры при любом изменении кода.

ygorshenin commented on map/benchmark_tool/benchmark_tool.pro:
```diff
@@ -6,6 +6,7 @@ CONFIG -= app_bundle
 TEMPLATE = app
 
 ROOT_DIR = ../..
+# TODO(AlexZ): Remove map dependency.
```
If benchmark tool should not depend on map probably it's worth to move it out from map directory?

vng replied:
Не вижу тут никакой проблемы. Зачем удалять эту зависимость?

deathbaba replied:
To remove dependency, we need to change it’s code. And move only after that. It’s for separate PR and I put TODO for it.

deathbaba replied:
Посмотрел внимательнее - да, я ошибся, бенчмарк не может без map из-за использования FeaturesFetcher.

vng commented on omim.pro:
```diff
@@ -26,75 +29,140 @@ win32:CONFIG(drape) {
-  SUBDIRS += platform
-  SUBDIRS += stats
-  SUBDIRS += storage
+  SUBDIRS *= platform stats storage
```
у нас еще актуальна библиотека stats?

deathbaba replied:
Да, это наша статистика, которая на самом деле создает либу stats_client

deathbaba commented:
Removed controversial commit with dependency checks, it should be re-tested separately. Also updated speed test results, they are still great!

Please PTAL @Zverik @ygorshenin @syershov @vng

vng commented:
LGTM

igrechuhin commented:
ssd debug qmake iphone

real  12m9.225s
user  66m11.962s
sys 3m58.966s

ssd debug qmake iphone rebuild

real  0m22.935s
user  1m22.495s
sys 0m4.647s

deathbaba commented:
It is faster than without it, right?

Merge please.

> On 10.11.2015, at 04.51, Илья Гречухин <notifications@github.com> wrote:
> 
> ssd debug qmake iphone
> 
> real 12m9.225s
> user 66m11.962s
> sys 3m58.966s
> 
> ssd debug qmake iphone rebuild
> 
> real 0m22.935s
> user 1m22.495s
> sys 0m4.647s
> 
> —
> Reply to this email directly or view it on GitHub.
>

syershov commented:
LGTM

ygorshenin commented:
LGTM

igrechuhin commented:
LGTM

bykoianko commented:
LGTM @Zverik отался твой LGTM

Zverik commented:
Ладно, доверюсь вам. LGTM

Zverik commented:
Прошу у всех прощения. Увидел четыре LGTM и поверил на слово, а не стоило. Под линуксом сейчас вообще ничего не собирается, при этом ошибок сборки не пишет.

Zverik commented on omim.pro:
```diff
@@ -26,75 +29,140 @@ win32:CONFIG(drape) {
+  CONFIG(desktop) {
+    benchmark_tool.subdir = map/benchmark_tool
+    benchmark_tool.depends = 3party base coding geometry platform indexer map
+    SUBDIRS *= benchmark_tool
```
Почему мы собираем benchmark_tool в десктопе безусловно? Это не тест? Его всегда хотят те, кто собирает десктоп?

deathbaba replied:
It is not a test.

> On 11.11.2015, at 04.38, Ilya Zverev <notifications@github.com> wrote:
> 
> In omim.pro:
> 
> >      }
> >    }
> >  
> > +  CONFIG(desktop) {
> > +    benchmark_tool.subdir = map/benchmark_tool
> > +    benchmark_tool.depends = 3party base coding geometry platform indexer map
> > +    SUBDIRS *= benchmark_tool
> Почему мы собираем benchmark_tool в десктопе безусловно? Это не тест? Его всегда хотят те, кто собирает десктоп?
> 
> —
> Reply to this email directly or view it on GitHub.
>
